pkg/inject/inject.go

```diff
@@ -187,15 +188,41 @@ func (conf *ResourceConfig) GetPatch(
 		}
 
 		report.update(conf)
+		if !shouldInject(conf, report) && !shouldOverrideConfig(conf) {
+			return &Patch{}, []Report{report}, nil
+		}
+
+		// populate the proxy's configurable properties with either overrides or
+		// defaults
+		proxy := conf.setProxyConfigs(identity)
```

Comment:
Say `shouldInject()` returned false because `HostNetwork` is true, yet the user added config annotations unknowingly. This will inject the proxy when it shouldn't. I think `shouldInject()` needs to be split into different methods to be able to distinguish  when pods should not be injected at all, and when they are already injected.

Reply:
This might also help clarifying a bit the logic to decide when to inject/override inside `GetPatch()`, which is currently not obvious IMO.

Reply:
Yes, definitely a bug in the `if` statement. As a matter of fact, it breaks `linkerd inject`, because the received YAML will already have a proxy spec, and if config annotations are present, the bug you described will occur, leading to some downstream complication in the YAML (i.e. the CLI calls `GetPatch()` once, and then the webhook calls it again when the deployment is created.) :boom:

I'm not sure how splitting up the `shouldInject()` will help, as I imagine it will require some clever ways to chain these different methods together, to satisfy both the webhook and CLI code path.

What I ended up doing was to perform the `shouldInject()` checks earlier in both the CLI and webhhook. So if `shouldInject()` failed, we won't even call the `GetPatch()` function. This also helps to keep `GetPatch()` to focus only on generating JSON patches, and the decision on whether the configs need to be overridden.

The commit is at dde345f75aeeff75712406c39804e0c53705efb7 This seems to work for the webhook, but I'm now battling with how the CLI reports unit tests. Maybe you can take a look at that commit to see if it makes sense? Thanks!

Reply:
> What I ended up doing was to perform the shouldInject() checks earlier in both the CLI and webhhook. So if shouldInject() failed, we won't even call the GetPatch() function. This also helps to keep GetPatch() to focus only on generating JSON patches, and the decision on whether the configs need to be overridden.

Yes, this looks much better to me :+1: 
Note though that your `shouldInject()` functions will return false whenever there's already a proxy attached because `healthcheck.HasExistingSidecars()` doesn't distinguish between linkerd proxies and 3rd party proxies, so this will prevent config overriding. That should be an easy fix :wink: